Late and edit

pdx-cs-lbryan opened the pull request:
This branch will be used for changes to reviews including:
New features:
* marking reviews as late -- status: done
* allowing students to select a capstone session for reviews -- status: done
* allow students to edit reviews -- status: done

Fixes:
* Fix: checking review state before student -- status: done
* Fix: scores cannot be negative -- status: done

Clean up for review: done
Testing: tested all old and new features

Testing Guide:
In addition to general testing, there are three main features to test:

**Late Reviews Are Appropriately Marked as Such (and On Time Reviews are Not)**:
To test this feature, go through the normal processes of adding yourself for review, setting review reviewing your team, submitting your review, etc.

Check the database to see if your reviews are being marked on time or as late
Check the professor view of student reports for if late reviews are noted as such, and if on time reviews are noted as not late

A review should be late if it's submitted after the end date for the review of its type

You should check:
Midterms: on time and late
Finals: on time and late
(Note: editing reviews does not change if they are late or not)

**Capstone session selection:**
Reviews now take into account multiple capstone sessions for a student, and behave according to the selected sessions terms.

Add yourself to the number of capstone sessions you wish to test, as a professor

Then navigate to the student page

Do you see a dropdown selection menu?
Are all your capstone sessions displayed there?
---you should test one and several sessions

Does clicking 'review' take you to the correct capstone session?
Do you see the appropriate teammates?
Is the review closed (before dates), midterm (empty fields, first time), midterm (populated data, editing), final (first time, empty fields), final (populated data, editing) appropriately?
When you submit a review, are reviews submitted for the correct team and capstone session?
Is the correct student record updated for midterm/final done?'

**Editing Reviews**
To edit a review:
Go through the usual process of filling out a review
Now go back to the student dashboard
Select the same capstone session if not already selected
Click review

Are you back on the review page?
Are the data fields appropriately populated?
If you filled out a midterm review previously, but not final, and you go to final, do you see a blank review?

When you change data and submit, are the appropriate data entries changed (should only change the entries you are directly changing)?
Is the rest of the data kept the same?

Are you able to edit midterm and final reviews appropriately?

If a team member is removed, is the appropriate data still displayed for the remaining team members?

If you made a mistake, like entering the wrong points total, is data incorrectly overwritten?

pdx-cs-lbryan committed:
disallow a value less than 0 for points on reviews

pdx-cs-lbryan committed:
add model support for checking if reviews are late

pdx-cs-lbryan committed:
add error handling for checking late reviews

pdx-cs-lbryan committed:
add support for checking if reviews are late upon submission

pdx-cs-lbryan committed:
Merge branch 'master' of https://github.com/capstoneteamb/capstone360 into late-and-edit

pdx-cs-lbryan committed:
start dashboard changes

pdx-cs-lbryan committed:
allow front-end support for students selecting capstone session

pdx-cs-lbryan committed:
add backend support for selecting capstone session to review

pdx-cs-lbryan committed:
fix post and marking correct student record

pdx-cs-lbryan committed:
fixes for marking student records upon review submission

pdx-cs-lbryan committed:
display tech_mastery info retrieved on review

pdx-cs-lbryan committed:
all radio buttons retain entered values when editing

pdx-cs-lbryan committed:
retain points, strengths, weak, and traits for editing reviews

pdx-cs-lbryan committed:
retain learned and proud fields for editing

pdx-cs-lbryan committed:
Merge branch 'master' of https://github.com/capstoneteamb/capstone360 into late-and-edit

pdx-cs-lbryan committed:
add editing for submitted reviews

pdx-cs-lbryan committed:
clean up for new review features

pdx-cs-lbryan committed:
flake8 fixes

pdx-cs-lbryan committed:
more flake8 fixes

pdx-cs-lbryan committed:
remove whitespace from line for flake8

pdx-cs-lbryan committed:
one last flake8 fix

pdx-cs-lbryan committed:
add error handling for get data

necrobuffalo reviewed (changes requested):
Session selection and late reviews look good, didn't run into any issues until I got to editing reviews.
* Radio buttons populate automatically with old data, but text boxes don't.
* When trying to edit a late final review, on submit, I receive the "You have no open reviews" error and my edits don't save. If I adjust the dates again so it's within the window, trying to edit the final review causes a 500 error (no traceback sadly) and prints this to the console:
```
/home/adam/src/capstone360/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py:551: SAWarning: Session's state has been changed on a non-active transaction - this state will be discarded.
  "Session's state has been changed on "
Submission Errror
```

I have a few style complaints as well:
* The name `cap` is unclear - if I hadn't been testing these changes, I would not have any idea what it is or refers to.
* I'm not going to be too harsh on this since I know you've been doing this style for a while but please put new class/function/module descriptions in docstrings and not comments. You can look at PEP 257 if you want an example of some well formatted ones.

pdx-cs-lbryan committed:
resolve merge conflict

pdx-cs-lbryan committed:
add docstrings, bugfix for model

pdx-cs-lbryan committed:
rename cap variable, add print statements for logging post actions

pdx-cs-lbryan committed:
add docstrings to form.py

pdx-cs-lbryan committed:
Add further logging statements

pdx-cs-lbryan committed:
bug fix, linting fixes

pdx-cs-lbryan committed:
Remove unnecessary import, revise review template

pdx-cs-lbryan committed:
strip whitetext from input on backend

pdx-cs-lbryan committed:
trim textarea input

pdx-cs-lbryan committed:
fix flake8 issues

pdx-cs-lbryan commented:
> Radio buttons populate automatically with old data, but text boxes don't.

I've had trouble reproducing this error across different hardware, browsers, and operating systems. I believe Jacob tested this out too and didn't report an error, so I'm not sure what's going on. I've taken the following steps:
* When adding info from textareas to the database, extra whitespace is now stripped
* some templating changes
* when a student begins to edit a review, the console now logs the object containing the data to populate the form.

If you still encounter this bug, could you provide:
-the console log of that object
-the html file that is produced

> When trying to edit a late final review, on submit, I receive the "You have no open reviews" error and my edits don't save. If I adjust the dates again so it's within the window, trying to edit the final review causes a 500 error (no traceback sadly) and prints this to the console...

Again here I've had trouble reproducing this bug and Jacob did not indicate having it either. If you still encounter the issue, I'd appreciate having as much info as possible about the steps taken prior to receiving this issue.

I've added some print statements to log the actions taken during a post request.

> The name cap is unclear - if I hadn't been testing these changes, I would not have any idea what it is or refers to.

Changed 

> I'm not going to be too harsh on this since I know you've been doing this style for a while but please put new class/function/module descriptions in docstrings and not comments. You can look at PEP 257 if you want an example of some well formatted ones.

I believe all my functions/classes now use docstrings

necrobuffalo committed:
Don't log PII

Don't need this for bugfixing anymore.

necrobuffalo reviewed (approved):
Well, whatever was going on on my machine it's fine now. I edited a bit so we're not logging PII (don't want student names to end up in plaintext in the apache logs, etc.) and will merge.
